core/store/src/flat_state.rs

```diff
@@ -667,86 +677,74 @@ pub mod store_helper {
-        catchup_status_key
-    }
-
-    fn get_catchup_status(store: &Store, shard_id: ShardId) -> bool {
-        let status: Option<bool> =
-            store.get_ser(crate::DBCol::FlatStateMisc, &catchup_status_key(shard_id)).expect(
-                format!("Error reading catchup status for flat state for shard {shard_id}")
-                    .as_str(),
-            );
-        match status {
-            None => false,
-            Some(status) => {
-                assert!(
-                    status,
-                    "Catchup status for flat state for shard {} must be true if stored",
-                    shard_id
-                );
-                true
+        let value = match status {
+            FlatStorageCreationStatus::FetchingState(status) => {
+                let mut value = vec![FETCHING_STATE];
+                value.extend_from_slice(&status.try_to_vec().unwrap());
+                value
             }
-        }
-    }
-
-    pub fn start_catchup(store_update: &mut StoreUpdate, shard_id: ShardId) {
-        store_update
-            .set_ser(crate::DBCol::FlatStateMisc, &catchup_status_key(shard_id), &true)
-            .expect(format!("Error setting catchup status for shard {shard_id}").as_str());
-    }
-
-    pub fn finish_catchup(store_update: &mut StoreUpdate, shard_id: ShardId) {
-        store_update.delete(crate::DBCol::FlatStateMisc, &catchup_status_key(shard_id));
+            FlatStorageCreationStatus::CatchingUp(block_hash) => {
+                let mut value = vec![CATCHING_UP];
+                value.extend_from_slice(block_hash.as_bytes());
+                value
+            }
+            status @ _ => {
+                panic!("Attempted to write incorrect flat storage creation status {status:?} for shard {shard_id}");
+            }
+        };
+        store_update.set(crate::DBCol::FlatStateMisc, &creation_status_key(shard_id), &value);
     }
 
     pub fn get_flat_storage_creation_status(
         store: &Store,
         shard_id: ShardId,
     ) -> FlatStorageCreationStatus {
         match get_flat_head(store, shard_id) {
-            None => FlatStorageCreationStatus::SavingDeltas,
             Some(_) => {
-                if let Some(fetching_state_status) = get_fetching_state_status(store, shard_id) {
-                    FlatStorageCreationStatus::FetchingState(fetching_state_status)
-                } else if get_catchup_status(store, shard_id) {
-                    FlatStorageCreationStatus::CatchingUp
-                } else {
-                    FlatStorageCreationStatus::Ready
+                return FlatStorageCreationStatus::Ready;
+            }
+            None => {}
+        }
+
+        let value = store
+            .get(crate::DBCol::FlatStateMisc, &creation_status_key(shard_id))
+            .expect("Error reading status from storage");
+        match value {
+            None => FlatStorageCreationStatus::SavingDeltas,
```

Comment:
Some possible improvements but I think the current way is fine too. Is it possible to also explicitly write "FlatStorageCreationStatus::SavingDelta" to the column instead of based on the assumption that it if the value doesn't exist then flat storage is saving delta? I think it will be more bug prone for future changes, for example, if you add another enum value here.

Reply:
I think it's possible to do a "migrate_34_to_35" for DB version, but I need to check it separately.